src/compiler/program.ts

```diff
@@ -1146,8 +1146,10 @@ namespace ts {
                 const typeChecker = getDiagnosticsProducingTypeChecker();
 
                 Debug.assert(!!sourceFile.bindDiagnostics);
-                // For JavaScript files, we don't want to report semantic errors unless explicitly requested.
-                const includeBindAndCheckDiagnostics = !isSourceFileJavaScript(sourceFile) || isCheckJsEnabledForFile(sourceFile, options);
+
+                // By default, only type-check .ts, .tsx, and 'External' files (external files are added by plugins)
+                const includeBindAndCheckDiagnostics = sourceFile.scriptKind === ScriptKind.TS || sourceFile.scriptKind === ScriptKind.TSX ||
+                    sourceFile.scriptKind === ScriptKind.External || isCheckJsEnabledForFile(sourceFile, options);

```

Comment:
Does something about our parsing prevent `checkJsDirective` from being set for non-JS files (e.g. for JSON files)?

Reply:
Wouldn't that fall through to the `isCheckJsEnabledForFile(sourceFile, options)` at the end of the OR clauses if so?

Reply:
Yes, wouldn't that be bad?